File: java/2-finish/akka-streams/src/main/java/pipelines/workshop/processor/SanctionedCountriesFilter.java
package pipelines.workshop.processor;

import akka.NotUsed;
import akka.stream.javadsl.FlowWithContext;
import pipelines.akkastream.AkkaStreamlet;
import pipelines.akkastream.PipelinesContext;
import pipelines.akkastream.javadsl.util.Either;
import pipelines.akkastream.util.javadsl.SplitterLogic;
import pipelines.streamlets.StreamletShape;
import pipelines.streamlets.avro.AvroInlet;
import pipelines.streamlets.avro.AvroOutlet;
import pipelines.workshop.schema.java.CardPayment;

public class SanctionedCountriesFilter extends AkkaStreamlet {

    AvroInlet<CardPayment> inlet = AvroInlet.<CardPayment>create("in", CardPayment.class);
    AvroOutlet<CardPayment> manualOutlet = AvroOutlet.<CardPayment>create("manual", CardPayment::countryCode, CardPayment.class);
    AvroOutlet<CardPayment> automaticOutlet = AvroOutlet.<CardPayment>create("auto", CardPayment::customerId, CardPayment.class);

    @Override
    public StreamletShape shape() {
        return StreamletShape.createWithInlets(inlet).withOutlets(manualOutlet, automaticOutlet);
    }

    @Override
    public SplitterLogic createLogic() {
        return new SplitterLogic<CardPayment, CardPayment, CardPayment>(inlet, manualOutlet, automaticOutlet, getStreamletContext()) {
            @Override
            public FlowWithContext<CardPayment, PipelinesContext, Either<CardPayment, CardPayment>, PipelinesContext, NotUsed> createFlow() {
                return createFlowWithPipelinesContext().map(this::checkForSanctions);
            }

            Either<CardPayment, CardPayment> checkForSanctions(CardPayment payment) {
                if (payment.countryCode().equals("RU")) {
                    return Either.left(payment);
                } else {
                    return Either.right(payment);
                }
            }
        };
    }

//    @Override
//    public StreamletLogic createLogic() {
//        return new RunnableGraphStreamletLogic(getStreamletContext()) {
//
//            public FlowWithContext<CardPayment, PipelinesContext, CardPayment, PipelinesContext, NotUsed> createFlow() {
//                return FlowWithContext.<CardPayment, PipelinesContext>create().filter(payment -> payment.getCountryCode().equals("RU"));
//            }
//
//            public RunnableGraph<NotUsed> createRunnableGraph() {
//                return getAtLeastOnceSource(inlet).via(createFlow()).to(getAtLeastOnceSink(outlet));
//            }
//        };
//    }
}

File: scala/2-finish/akka-streams/src/main/scala/pipelines/workshop/processor/SanctionedCountriesFilter.scala
package pipelines.workshop.processor

import akka.stream.scaladsl.RunnableGraph
import pipelines.akkastream.scaladsl.{ FlowWithPipelinesContext, RunnableGraphStreamletLogic }
import pipelines.akkastream.{ AkkaStreamlet, StreamletLogic }
import pipelines.streamlets.avro.{ AvroInlet, AvroOutlet }
import pipelines.streamlets.{ StreamletShape, StringConfigParameter }
import pipelines.workshop.schema.CardPayment

class SanctionedCountriesFilter extends AkkaStreamlet {

  val inlet = AvroInlet[CardPayment]("in")
  val outlet = AvroOutlet[CardPayment]("out", _.countryCode)

  val shape = StreamletShape(inlet, outlet)

  val FilteringResource = StringConfigParameter(
    key = "sanctioned-countries",
    description = "Payments from these countries require manual processing",
    Some("RU")
  )

  override protected def createLogic(): StreamletLogic = new RunnableGraphStreamletLogic() {

    println(streamletConfig.entrySet())

    val sanctionedCountriesConfig = if (streamletConfig.hasPath(FilteringResource.key)) streamletConfig.getString(FilteringResource.key) else FilteringResource.defaultValue.get
    val sanctionedCountries = sanctionedCountriesConfig.split(",")

    def flow =
      FlowWithPipelinesContext[CardPayment].filter { payment ⇒ sanctionedCountries.contains(payment.countryCode) }

    override def runnableGraph(): RunnableGraph[_] =
      atLeastOnceSource(inlet).via(flow).to(atLeastOnceSink(outlet))
  }
}
